### ISSUE TITLE
hilbert transform filters from remez() have wrong phase

### ISSUE THREAD

**Issue by aholtzma** [scipy/scipy#7588]
The hilbert transform filters produced by remez() have the wrong phase. The positive frequencies have a +90 degree shift instead of a -90 degree shift (and vice versa for the negative frequencies).

Looking at the impulse response, you can see it is off by 180 degrees. The trailing edge of the hilbert impulse response should be positive, not negative.

![image](https://user-images.githubusercontent.com/409218/28075055-43c61d84-6628-11e7-8835-166b28277001.png)

The phase slope we get at zero makes intuitive sense - as we go from negative  frequencies to postive, we traverse the zero on the unit circle at +1. This will cause the phase from flip from -90 to +90. To fix this, when the filter type is 'hilbert', a scaling factor of -1 should be applied to the impulse response.
### Reproducing code example:
```
import scipy.signal
import matplotlib.pyplot as plt

h = scipy.signal.remez(29, [0.1, 0.9], [1], Hz=2, type='hilbert')
plt.stem(h)
plt.show()

```

### Scipy/Numpy/Python version information:
```
0.19.0 1.13.0 sys.version_info(major=3, minor=5, micro=2, releaselevel='final', serial=0)
```

**Comment by rgommers** [comment]
@charris IIRC you were our ``remez`` expert. Is this a bug or a matter of different definitions? And if the former, can we just change the behavior?

**Comment by lucascolley** [comment]
hey @DietBru, maybe you have some time to look at this?

**Comment by DietBru** [comment]
I agree that multiplying the impulse response with -1 would produce the expected result (I also crosschecked with Octave). 
I do have bit of trouble comprehending the corresponding [C-Code](https://github.com/scipy/scipy/blob/ff99a9b900785c71bd049ff72f7eeecd9c08d31c/scipy/signal/_sigtoolsmodule.c#L565) and I am not comfortable modifying it.

Furthermore, this does not seem to be the only problem with `remez()`, since I am able to reproduce  the crash reported in #11968 in SciPy 1.12 with
```python
from scipy.signal import remez
remez(10, [0,0.01], [1], type='differentiator')
```
Perhaps time would be better spent, reimplementing the function in pure Python (or Cython) to make it easier to maintain. At the moment, availble time is too scarce to tackle this myself, but I would be happy to mentor...

**Comment by ev-br** [comment]
Do you have a reference off which to implement it? I'll need a cuda-compatible version for cupy anyways.

**Comment by DietBru** [comment]
You could take a look at this [paper](https://inria.hal.science/hal-01136005v3/document) titled "A robust and scalable implementation of the Parks-McClellan algorithm for designing FIR filters" from Silviu-Ioan Filip as a starting point.

**Comment by ja3443** [comment]
The algorithm in _sigtoolsmodule.c appears to be based on the original Fortran code by McClellan, Parks and Rabiner [1]. This code is available at various locations, e.g. [2], if you want to play with it (it works with gfortran if -std=legacy is set, also the i1mach function is not found but it can be replaced by setting input=5 and iout=6).

The bug for type hilbert seems to be present in the original code but further investigation would be needed to confirm that. For type differentiator, which is somewhat related, the results are correct.

It looks like a fix could be added to the code in lines 742-748 of _sigtoolsmodule.c to change the coefficients for type hilbert. The variable jtype contains the filter type (jtype=3 is hilbert).

Here is a test program in python. It can be used to test both differentiator and hilbert results. If the fix is correct the differentiator results will not change and the hilbert results will change such that the impulse response (h) is inverted and the phase (angle) is inverted as compared to the current results. 
[test-remez.py.txt](https://github.com/scipy/scipy/files/14627933/test-remez.py.txt)

[1] A computer program for designing optimum FIR linear phase digital filters, J. McClellan, T. Parks, L. Rabiner, IEEE Transactions on Audio and Electroacoustics, Dec. 1973. https://ieeexplore.ieee.org/abstract/document/1162525
[2] https://michaelgellis.tripod.com/dsp/pgm21.html

**Comment by ev-br** [comment]
That code is GPLed, sadly.

**Comment by ev-br** [comment]
@sfilip responded with an agreement to re-license his code for SciPy.  Thank you @sfilip!

**Comment by sfilip** [comment]
Hi! My pleasure. I'm thrilled to help in any way that I can. As @ev-br mentioned, I have relicensed the code to BSD.

A first important matter to discuss is what kind of dependencies are allowed. There are two main ones in `firpm`, but I think they can probably be removed/replaced.

1. [eigen](https://gitlab.com/libeigen/eigen): I use the eigenvalue solver in Eigen to compute the set of points that will form the basis of the next iteration of the Remez/Parks-McClellan algorithm. Since my code started as a research project looking into the numerics of the algorithm, I wanted an eigensolver that was flexible enough to work with multiple precision floating-point arithmetic (this is why the routines in `firpm` are templated to work with `double`, `long double` and `mpfr` multiple precision arithmetic). Eigen offers that, but I would imagine that in SciPy only a double precision version of the algorithm has its place. If that is the case, then relying on eigen might not be a necessity, since SciPy already has an eigenvalue problem solver in ```scipy.linalg.eig``` that works with double precision matrices. 
2. OpenMP: I use OpenMP to parallelize the search for the aforementioned set of points. This can be disabled, for instance.

Another option is to write a pure Python version. I would go for this if the integration efforts on the C++ version look too cumbersome.

In any eventuality, I am happy to help with the integration or re-write of a Python version.

**Comment by lucascolley** [comment]
Great to hear @sfilip ! I'm going to create a new issue for discussion of this integration / rewrite, to keep this one about the original bug report. Will ping and x-ref.
